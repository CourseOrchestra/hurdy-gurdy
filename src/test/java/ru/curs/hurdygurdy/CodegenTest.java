package ru.curs.hurdygurdy;

import com.squareup.javapoet.TypeSpec;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Comparator;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class CodegenTest {
    private JavaCodegen codegen = new JavaCodegen(
            GeneratorParams.rootPackage("com.example").generateResponseParameter(true));
    Path result;

    @BeforeEach
    void setUp() throws IOException {
        result = Files.createTempDirectory("codegen");
    }

    @AfterEach
    void tearDown() throws IOException {
        Files.walk(result)
                .sorted(Comparator.reverseOrder())
                .map(Path::toFile)
                .forEach(File::delete);
    }

    @Test
    void generateSample1() throws IOException {
        codegen.generate(Path.of("src/test/resources/sample1.yaml"), result);
        Approvals.verify(getContent(result));
    }

    @Test
    void generateSample2() throws IOException {
        codegen.generate(Path.of("src/test/resources/sample2.yaml"), result);
        Approvals.verify(getContent(result));
    }

    @Test
    void doNotGenerateResponseParameter() throws IOException {
        codegen = new JavaCodegen(GeneratorParams.rootPackage("com.example")
                .generateResponseParameter(false)
                .generateApiInterface(true));
        codegen.generate(Path.of("src/test/resources/sample1.yaml"), result);
        Approvals.verify(getContent(result));
    }

    @Test
    void generateCommonParameters() throws IOException {
        codegen.generate(Path.of("src/test/resources/commonparam.yaml"), result);
        Approvals.verify(getContent(result));
    }

    @Test
    void generateMultipart() throws IOException {
        codegen.generate(Path.of("src/test/resources/multipart.yaml"), result);
        Approvals.verify(getContent(result));
    }

    @Test
    void generateFromSpecs() throws IOException {
        codegen.addTypeSpec(ClassCategory.DTO, TypeSpec.interfaceBuilder("Intf1").build());
        codegen.addTypeSpec(ClassCategory.DTO, TypeSpec.interfaceBuilder("Intf2").build());
        codegen.addTypeSpec(ClassCategory.CONTROLLER, TypeSpec.interfaceBuilder("Intf3").build());
        codegen.generate(result);
        Approvals.verify(getContent(result));
    }

    @Test
    void paramsOverriding() throws IOException {
        codegen.generate(Path.of("src/test/resources/twoparams.yaml"), result);
        Approvals.verify(getContent(result));
    }

    @Test
    void camelCase() throws IOException {
        codegen = new JavaCodegen(GeneratorParams.rootPackage("com.example")
                .forceSnakeCaseForProperties(false));
        codegen.generate(Path.of("src/test/resources/camelcase.yaml"), result);
        Approvals.verify(getContent(result));
    }

    @Test
    void dictionarySupport() throws IOException {
        codegen.generate(Path.of("src/test/resources/dictionary.yaml"), result);
        Approvals.verify(getContent(result));
    }

    String getContent(Path path) throws IOException {
        return Files.walk(path)
                .sorted()
                .flatMap(p -> Stream.concat(
                        Stream.of(
                                String.format("---%n"),
                                String.format("%s%n", p.toString()
                                        .replaceAll(String.format("\\%s", File.separator), "/")
                                        .substring(result.toString().length()))
                        ),
                        readFile(p))
                ).collect(Collectors.joining());
    }

    Stream<String> readFile(Path path) {
        String result;
        if (Files.isReadable(path)) {
            try {
                result = Files.readString(path);
            } catch (IOException e) {
                result = null;
            }
            return Stream.ofNullable(result);
        } else {
            return Stream.empty();
        }
    }
}