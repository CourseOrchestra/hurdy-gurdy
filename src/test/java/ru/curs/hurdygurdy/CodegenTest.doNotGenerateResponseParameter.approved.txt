---

---
/com
---
/com/example
---
/com/example/controller
---
/com/example/controller/Controller.java
package com.example.controller;

import com.example.dto.MenuDTO;
import java.lang.Boolean;
import java.lang.Integer;
import java.lang.String;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;

interface Controller {
  @GetMapping(
      value = "/menu",
      produces = "application/json"
  )
  MenuDTO getMenu();

  @DeleteMapping("/menu")
  void deleteMenu();

  @GetMapping(
      value = "/api/v1/affiliate/{id}/playeractivities/csv",
      produces = "text/csv"
  )
  String affiliatePlayerActivitiesCSV(@PathVariable(name = "id") int id,
      @RequestParam(required = false, name = "status") Boolean status);

  @GetMapping(
      value = "/api/v1/bills",
      produces = "application/json"
  )
  int getBills(@RequestParam(required = false, name = "affiliate_id") Integer affiliateId);
}
---
/com/example/dto
---
/com/example/dto/MenuDTO.java
package com.example.dto;

import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import java.util.List;
import lombok.Data;

@Data
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public class MenuDTO {
  private List<MenuItemDTO> menuItems;
}
---
/com/example/dto/MenuItemDTO.java
package com.example.dto;

import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import java.lang.String;
import java.util.List;
import lombok.Data;

@Data
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public class MenuItemDTO {
  private String id;

  private boolean enabled;

  private String name;

  private List<MenuItemDTO> menuItems;
}
